LifeCycleAssessment_Engine/Compute/EvaluateEnvironmentalProductDeclarationByMass.cs

```diff
@@ -44,40 +44,45 @@ public static partial class Compute
+            List<double> gwpByMaterial = new List<double>();
+            List<double> volumeByRatio = elementM.IMaterialComposition().Ratios.Select(x => volume * x).ToList();
+            List<double> densityOfMassEpd = Query.GetElementDensity(elementM);
+            List<double> massOfObj = new List<double>();
+
+            if (densityOfMassEpd == null)
             {
-                BH.Engine.Reflection.Compute.RecordError("This EnvironmentalProductDeclaration's QuantityType is not Mass. Please supply a Mass-based EPD or try a different method.");
+                BH.Engine.Reflection.Compute.RecordError("Density could not be found. Material density is required for all objects using Mass-based evaluations.");
                 return null;
             }
-            else
-            {
-                double mass = elementM.Mass();
-                List<double> epdVal = elementM.GetEvaluationValue(field);
-                List<double> massByRatio = elementM.IMaterialComposition().Ratios.Select(x => mass * x).ToList();
-                List<double> gwpByMaterial = new List<double>();
 
-                for (int x = 0; x < epdVal.Count; x++)
-                    gwpByMaterial.Add(epdVal[x] * massByRatio[x]);
-
-                if (epdVal.Sum() <= 0 || epdVal == null)
-                {
-                    BH.Engine.Reflection.Compute.RecordError($"No value for {field} can be found within the supplied EPD.");
-                    return null;
-                }
+            for (int x = 0; x < volumeByRatio.Count; x++)
+            {
+                massOfObj.Add(volumeByRatio[x] * densityOfMassEpd[x]);
+            }
 
-                if (mass <= 0 || mass == double.NaN)
+            for (int x = 0; x < epdVal.Count; x++)
+            {
+                if (double.IsNaN(epdVal[x]))
+                    gwpByMaterial.Add(double.NaN);
+                else
                 {
-                    BH.Engine.Reflection.Compute.RecordError("Mass cannot be calculated from object " + ((IBHoMObject)elementM).BHoM_Guid);
-                    return null;
+                    gwpByMaterial.Add(epdVal[x] * massOfObj[x]);
                 }
+            }
 
-                double quantity = gwpByMaterial.Sum();
+            if (epdVal == null || epdVal.Where(x => !double.IsNaN(x)).Sum() <= 0)
+            {
+                BH.Engine.Reflection.Compute.RecordError($"No value for {field} can be found within the supplied EPD.");
+                return null;
+            }
 
-                return new GlobalWarmingPotentialResult(((IBHoMObject)elementM).BHoM_Guid, field, 0, ObjectScope.Undefined, ObjectCategory.Undefined, ((IBHoMObject)elementM).GetAllFragments().Where(y => typeof(IEnvironmentalProductDeclarationData).IsAssignableFrom(y.GetType())).Select(z => z as IEnvironmentalProductDeclarationData).FirstOrDefault(), quantity);
+            double quantity = gwpByMaterial.Where(x => !double.IsNaN(x)).Sum();
```

Comment:
this is reflected across all methods to make sure the methods are only working on the relevant EPDs per QT